File: app/src/main/java/com/cliniva/enventory/ui/customer/CustomerFragment.java
package com.cliniva.enventory.ui.customer;


import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.cliniva.enventory.R;
import com.cliniva.enventory.adapter.RecyclerViewAdapter;
import com.cliniva.enventory.adapter.base.BaseRecyclerViewHolder;
import com.cliniva.enventory.event.OnCustomerItemClickedListener;
import com.cliniva.enventory.model.Customer;
import com.cliniva.enventory.ui.base.BaseFragment;
import com.cliniva.enventory.viewholder.CustomerHolder;

import java.util.List;

/**
 * A simple {@link Fragment} subclass.
 */
public class CustomerFragment extends BaseFragment implements CustomerContract.View, OnCustomerItemClickedListener {

    private RecyclerView mCustomerList;
    private CustomerContract.Presenter mCustomerPresenter;

    public CustomerFragment() {
        // Required empty public constructor
    }

    public static CustomerFragment getInstance(){
        CustomerFragment fragment = new CustomerFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public int getLayoutRes() {
        return R.layout.fragment_customer;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mCustomerPresenter = new CustomerPresenter(this);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        setToolbar(view, mCustomerPresenter.getContext().getString(R.string.action_customer), false);

        mCustomerList = view.findViewById(R.id.rv_customer_list);
        LinearLayoutManager layoutManager = new LinearLayoutManager(mCustomerPresenter.getContext());
        mCustomerList.setLayoutManager(layoutManager);
        mCustomerPresenter.onLoadList();
    }

    @Override
    public void setListToView(List<Customer> customerList) {
        RecyclerViewAdapter<Customer, OnCustomerItemClickedListener> mCustomerAdapter = new RecyclerViewAdapter<Customer, OnCustomerItemClickedListener>(customerList) {
            @NonNull
            @Override
            public BaseRecyclerViewHolder<Customer, OnCustomerItemClickedListener> onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
                ViewDataBinding binding = DataBindingUtil.inflate(LayoutInflater.from(mCustomerPresenter.getContext()), R.layout.item_customer, parent, false);
                return new CustomerHolder(binding);
            }
        };
        mCustomerAdapter.setListener(this);
        mCustomerList.setAdapter(mCustomerAdapter);
    }

    @Override
    public void onSingleViewClicked(View v, int position) {

    }

    @Override
    public void onItemClickListener(Customer item, int position) {

    }
}

File: app/src/main/java/com/amirsons/inventory/ui/fragment/customer/CustomerFragment.kt
package com.amirsons.inventory.ui.fragment.customer


import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import com.amirsons.inventory.R
import com.amirsons.inventory.datamanager.model.Customer
import com.amirsons.inventory.event.OnCustomerItemClickedListener
import com.amirsons.inventory.ui.recyclerview.base.BaseRecyclerViewHolder
import com.amirsons.inventory.ui.recyclerview.base.RecyclerViewAdapter
import com.amirsons.inventory.ui.recyclerview.viewholder.CustomerHolder
import com.amirsons.inventory.ui.base.BaseFragment
import com.amirsons.inventory.ui.widgets.BottomSheetAddCustomer
import com.amirsons.inventory.ui.widgets.OnAddCustomerConfirm
import kotlinx.android.synthetic.main.fragment_customer.*
import java.util.*


/**
 * A simple [BaseFragment] subclass.
 */
class CustomerFragment : BaseFragment(), CustomerView {

    private lateinit var mCustomerPresenter: CustomerPresenter
    private lateinit var mCustomerAdapter: RecyclerViewAdapter<Customer, OnCustomerItemClickedListener>

    override val contentLayout: Int
        get() = R.layout.fragment_customer

    companion object {

        val instance: CustomerFragment
            get() {
                val fragment = CustomerFragment()
                val args = Bundle()
                fragment.arguments = args
                return fragment
            }
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mCustomerPresenter = CustomerMvp(this)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setToolbar(view, mCustomerPresenter.context.getString(R.string.action_customer), false)

        // init the customer list recyclerview adapter
        initCustomerListAdapter()

        // get all customer
        mCustomerPresenter.onLoadCustomerList()

        btn_fab.setOnClickListener {

           BottomSheetAddCustomer(context, "Add Customer Details" , object : OnAddCustomerConfirm {

                override fun onConfirm(item: Customer): Boolean {

                    // add new customer
                    mCustomerPresenter.onAddCustomerClick(item)

                    // return true to dismiss the dialog
                    return true
                }

            }).show()

        }
    }

    private fun initCustomerListAdapter() {

        mCustomerAdapter = object : RecyclerViewAdapter<Customer, OnCustomerItemClickedListener>() {

            override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseRecyclerViewHolder<Customer, OnCustomerItemClickedListener> {
                return CustomerHolder(inflate(R.layout.item_customer, parent))
            }
        }

        rv_customer_list.adapter = mCustomerAdapter
    }

    override fun setCustomerListToView(customerList: ArrayList<Customer>) {
        mCustomerAdapter.setItems(customerList)
    }
}
